src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,545 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Transactions;
+using NHibernate.AdoNet;
+using NHibernate.Engine;
+using NHibernate.Engine.Transaction;
+using NHibernate.Impl;
+using NHibernate.Util;
+
+namespace NHibernate.Transaction
+{
+	/// <summary>
+	/// <see cref="ITransaction"/> factory implementation supporting system
+	/// <see cref="System.Transactions.Transaction"/>.
+	/// </summary>
+	public class AdoNetWithSystemTransactionFactory : AdoNetTransactionFactory
+	{
+		private static readonly IInternalLogger _logger = LoggerProvider.LoggerFor(typeof(ITransactionFactory));
+
+		/// <summary>
+		/// See <see cref="Cfg.Environment.SystemTransactionCompletionLockTimeout"/>.
+		/// </summary>
+		protected int SystemTransactionCompletionLockTimeout { get; private set; }
+		/// <summary>
+		/// See <see cref="Cfg.Environment.UseConnectionOnSystemTransactionPrepare"/>.
+		/// </summary>
+		protected bool UseConnectionOnSystemTransactionPrepare { get; private set; }
+
+		/// <inheritdoc />
+		public override void Configure(IDictionary<string, string> props)
+		{
+			base.Configure(props);
+			SystemTransactionCompletionLockTimeout =
+				PropertiesHelper.GetInt32(Cfg.Environment.SystemTransactionCompletionLockTimeout, props, 5000);
+			if (SystemTransactionCompletionLockTimeout < -1)
+				throw new HibernateException(
+					$"Invalid {Cfg.Environment.SystemTransactionCompletionLockTimeout} value: {SystemTransactionCompletionLockTimeout}. It can not be less than -1.");
+			UseConnectionOnSystemTransactionPrepare =
+				PropertiesHelper.GetBoolean(Cfg.Environment.UseConnectionOnSystemTransactionPrepare, props, true);
+		}
+
+		/// <inheritdoc />
+		public override void EnlistInSystemTransactionIfNeeded(ISessionImplementor session)
+		{
+			// Handle the transaction on the originating session only.
+			var originatingSession = session.ConnectionManager.Session;
+
+			if (!originatingSession.ConnectionManager.ShouldAutoJoinTransaction)
+			{
+				return;
+			}
+
+			// Ensure the session does not run on a thread supposed to be blocked, waiting
+			// for transaction completion.
+			originatingSession.TransactionContext?.Wait();
+
+			var transaction = System.Transactions.Transaction.Current;
+			// We may have defined the transaction context before having the connection, so we
+			// need to ensure enlistment even when the transaction context is already defined.
+			// But avoid redefining to a sub-scope transaction.
+			if (originatingSession.TransactionContext == null ||
+				((SystemTransactionContext)originatingSession.TransactionContext).AmbientTransaction == transaction)
+			{
+				originatingSession.ConnectionManager.EnlistIfRequired(transaction);
+			}
```

Comment:
I have decided not to do that for avoiding an additional possible breaking change. The current code allows to ignore sub-scopes, when bot the session and the connection get enlisted in the main scope. This is current NHibernate behavior too. I have documented it that way in the new system transaction section.

Now there was an error case which is still there: if the session is enlisted in a scope before having its connection, then acquires its connection inside a sub-scope while having connection auto-enlistment (`Enlist!=false` in connection string), the connection get enlisted in the sub-scope, causing the session to be unusable after the scope. Here, having the test you think about would allow to fail earlier, which would be better.

Trying to support them would be likely overly complicated so I have not attempted it either. (Would likely requires to bound the connection to the transaction context, switch at each new sub scope connection and enlist a new transaction context, ... Or support `ConnectionReleaseMode.AfterStatement` as said in NH-4011 comments, mandates it to be used for such having sub-scope supported, but may still need to enlist in many context.)

Do you rather have that explicit throw? If we add it, maybe should it be done with an additional option allowing to ignore instead of throwing, for easing the upgrade path.